packages/block-editor/src/hooks/font-size.js

```diff
@@ -0,0 +1,151 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { hasBlockSupport } from '@wordpress/blocks';
+import { useSelect } from '@wordpress/data';
+import TokenList from '@wordpress/token-list';
+
+/**
+ * Internal dependencies
+ */
+import {
+	getFontSize,
+	getFontSizeClass,
+	getFontSizeObjectByValue,
+	FontSizePicker,
+} from '../components/font-sizes';
+import { cleanEmptyObject } from './utils';
+
+export const FONT_SIZE_SUPPORT_KEY = '__experimentalFontSize';
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `fontSize` and `fontWeight` attributes.
+ *
+ * @param  {Object} settings Original block settings
+ * @return {Object}          Filtered block settings
+ */
+function addAttributes( settings ) {
+	if ( ! hasBlockSupport( settings, FONT_SIZE_SUPPORT_KEY ) ) {
+		return settings;
+	}
+
+	// Allow blocks to specify a default value if needed.
+	if ( ! settings.attributes.fontSize ) {
+		Object.assign( settings.attributes, {
+			fontSize: {
+				type: 'string',
+			},
+		} );
+	}
+
+	return settings;
+}
+
+/**
+ * Override props assigned to save component to inject font size.
+ *
+ * @param  {Object} props      Additional props applied to save element
+ * @param  {Object} blockType  Block type
+ * @param  {Object} attributes Block attributes
+ * @return {Object}            Filtered props applied to save element
+ */
+function addSaveProps( props, blockType, attributes ) {
+	if ( ! hasBlockSupport( blockType, FONT_SIZE_SUPPORT_KEY ) ) {
+		return props;
+	}
+
+	// Use TokenList to dedupe classes.
+	const classes = new TokenList( props.className );
+	classes.add( getFontSizeClass( attributes.fontSize ) );
+	const newClassName = classes.value;
+	props.className = newClassName ? newClassName : undefined;
```

Comment:
Noting here that this has the side-effect of removing duplicated classes for old blocks that are updated. I can't think of any reason why this would be a problem but thought to highlight anyway before merging. cc @youknowriad 

I've testing this with old content and works nicely, so I'm going to land in a few hours.